php_threads.go

```diff
@@ -0,0 +1,90 @@
+package frankenphp
+
+// #include "frankenphp.h"
+import "C"
+import (
+	"fmt"
+	"sync"
+	"sync/atomic"
+)
+
+var (
+	phpThreads           []*phpThread
+	terminationWG        sync.WaitGroup
+	mainThreadShutdownWG sync.WaitGroup
+	threadsReadyWG       sync.WaitGroup
+	shutdownWG           sync.WaitGroup
+	done                 chan struct{}
+	threadsAreDone       atomic.Bool
+	threadsAreBooting    atomic.Bool
+)
+
+// reserve a fixed number of PHP threads on the go side
+func initPHPThreads(numThreads int) error {
+	threadsReadyWG = sync.WaitGroup{}
+	threadsAreDone.Store(false)
+	done = make(chan struct{})
+	phpThreads = make([]*phpThread, numThreads)
+	for i := 0; i < numThreads; i++ {
+		phpThreads[i] = &phpThread{threadIndex: i}
+	}
+	if err := startMainThread(numThreads); err != nil {
+		return err
+	}
+
+	// initialize all threads as inactive
+	threadsReadyWG.Add(len(phpThreads))
+	shutdownWG.Add(len(phpThreads))
+	threadsAreBooting.Store(true)
+
+	for _, thread := range phpThreads {
+		thread.setInactive()
+		if !C.frankenphp_new_php_thread(C.uintptr_t(thread.threadIndex)) {
+			panic(fmt.Sprintf("unable to create thread %d", thread.threadIndex))
+		}
+	}
+	threadsReadyWG.Wait()
+	threadsAreBooting.Store(false)
```

Comment:
To the above, using a rwlock:

```suggestion
	threadsAreBooting.Lock()

	for _, thread := range phpThreads {
		thread.setInactive()
		if !C.frankenphp_new_php_thread(C.uintptr_t(thread.threadIndex)) {
			panic(fmt.Sprintf("unable to create thread %d", thread.threadIndex))
		}
	}
	threadsReadyWG.Wait()
	threadsAreBooting.Unlock()
```

Reply:
Same as above, the `WaitGroup` here assures that the waiting goes 'both ways'. Locks will just cause waiting 'one-way' and make this function return before threads have started.

Reply:
This, as well is now done with thread states.